v2/internal/controllers/generic_controller.go

```diff
@@ -184,86 +192,69 @@ func register(
+	}
+	if ownershipResult != nil {
+		return *ownershipResult, nil
 	}
 
-	result, err := gr.Reconciler.Reconcile(ctx, log, gr.Recorder, metaObj)
-	if readyErr, ok := conditions.AsReadyConditionImpactingError(err); ok {
-		log.Error(readyErr, "Encountered error impacting Ready condition")
-		err = gr.WriteReadyConditionError(ctx, metaObj, readyErr)
+	var result ctrl.Result
+	if !metaObj.GetDeletionTimestamp().IsZero() {
+		result, err = gr.delete(ctx, log, metaObj)
+	} else {
+		result, err = gr.createOrUpdate(ctx, log, metaObj)
 	}
 
 	if err != nil {
-		log.Error(err, "error during reconcile")
-		// NotFound is a superfluous error as per https://github.com/kubernetes-sigs/controller-runtime/issues/377
-		// The correct handling is just to ignore it and we will get an event shortly with the updated version to patch
-		// We must also ignore conflict here because updating a resource that
-		// doesn't exist returns conflict unfortunately: https://github.com/kubernetes/kubernetes/issues/89985. This is OK
-		// to ignore because a conflict means either the resource has been deleted (in which case there's nothing to do) or
-		// it has been updated, in which case there's going to be a new event triggered for it and we can count this
-		// round of reconciliation as a success andait for the next event.
-		return ctrl.Result{}, reconcilers.IgnoreNotFoundAndConflict(err)
+		severity, err := gr.writeReadyConditionErrorOrDefault(ctx, log, metaObj, err)
+		if err != nil {
+			log.Error(err, "error during reconcile")
+			// NotFound is a superfluous error as per https://github.com/kubernetes-sigs/controller-runtime/issues/377
+			// The correct handling is just to ignore it and we will get an event shortly with the updated version to patch
+			// We must also ignore conflict here because updating a resource that
+			// doesn't exist returns conflict unfortunately: https://github.com/kubernetes/kubernetes/issues/89985. This is OK
+			// to ignore because a conflict means either the resource has been deleted (in which case there's nothing to do) or
+			// it has been updated, in which case there's going to be a new event triggered for it and we can count this
+			// round of reconciliation as a success and wait for the next event.
+			return ctrl.Result{}, kubeclient.IgnoreNotFoundAndConflict(err)
+		}
+		if severity == conditions.ConditionSeverityError {
+			// Severity error is fatal, return fast and block requeue
+			return ctrl.Result{}, nil
+		}
 	}
 
 	if (result == ctrl.Result{}) {
 		// If result is a success, ensure that we requeue for monitoring state in Azure
+		conditions.SetCondition(metaObj, gr.PositiveConditions.Ready.Succeeded(metaObj.GetGeneration()))
 		result = gr.makeSuccessResult()
 	}
 
+	// Write the object
+	err = gr.CommitUpdate(ctx, log, originalObj, metaObj)
+	if err != nil {
+		return ctrl.Result{}, kubeclient.IgnoreNotFoundAndConflict(err)
```

Comment:
should we return `nil, kubeclient.IgnoreNotFoundAndConflict(err)` here? Or is there a reason behind returning an empty `Result` ?

Reply:
The `reconcile.Reconciler` interface (provided by `controller-runtime`) which we are implementing has signature:
`Reconcile(context.Context, Request) (Result, error)`. So we can't actually return `nil` here.

There are basically 3 possible returns from this function:
- `ctrl.Result{}, nil`: The resource was reconciled successfully and the event which triggered this reconcile can be deleted from the event queue.
- `ctrl.Result{Requeue: true}, nil`: The resource is not done, requeue and try again later.
- `ctrl.Result{}, err`: There was an error. This is basically the same as `ctrl.Result{Requeue: true}, nil` except that the error is logged and the backoff before retry might be a bit different.

I've copied a comment from above to above this line too. It says: 
```
// NotFound is a superfluous error as per https://github.com/kubernetes-sigs/controller-runtime/issues/377
// The correct handling is just to ignore it and we will get an event shortly with the updated version to patch
// We must also ignore conflict here because updating a resource that
// doesn't exist returns conflict unfortunately: https://github.com/kubernetes/kubernetes/issues/89985. This is OK
// to ignore because a conflict means either the resource has been deleted (in which case there's nothing to do) or
// it has been updated, in which case there's going to be a new event triggered for it and we can count this
// round of reconciliation as a success and wait for the next event.
```